Angular: Add getWebpackConfig for angular 12.2.x & 13.x.x

ThibaudAV opened the pull request:
Issue: 
TRy to fix #16355  https://github.com/storybookjs/storybook/pull/16528

## What I did

New way to make storybook work with angular 13.

As the monorepo sotrybook is not compatible with the typescript version of angular 13 it is not possible to make imports. This creates errors at compilation time. 

Moreover too many versions of typescript and webpack are to be supported at the same time in a single project to support angular 6+.

To ignore all these problems this solution uses JavaScript and a check of the angular-cli version to use the right "angular-cli webpack config extractor".

It only works with the angular storybook builder
Currently the only way to test is with an external project. 


## How to test

- `npx -p @angular/cli ng new angular13`
- `npx sb@next init`

- add storybook builder to angular.json like : 
```ts
{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "angular13": {
      "projectType": "application",
      "schematics": {
          ...
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          ...
        },
        "serve": {
          ...
        },
        "extract-i18n": {
          ...
        },
        "test": {
          ...
        },
        "storybook": { // <-  storybook architect config to use sb builder
          "builder": "@storybook/angular:start-storybook",
          "options": {
            "tsConfig": ".storybook/tsconfig.json",
            "browserTarget": "angular13:build",
            "compodoc": true,
            "port": 9009
          }
        }
      }
    }
  },
  "defaultProject": "angular13"
}
```
- clone this branche.
- init and build storybook
- copy stroybook `app/angular/dist` to angular13 `node_modules/@storybook/angular`
- run `yarn ng run angular13:storybook`


My try project : https://github.com/ThibaudAV/sb-angular13


- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

--- 


WIP

- [x] fix builder test
- [x] clean and resolve FIXME
- [x] validate in angular 13 project
- [x] validate in angular 13 library
- [x] validate in nx

ThibaudAV committed:
feat(angular): add getWebpackConfig for angular 12.2.x & 13.x.x

only work with angular storybook builder

ThibaudAV commented on app/angular/src/server/angular-cli-webpack-13.x.x.js:
```diff
@@ -0,0 +1,83 @@
+const { filterOutStylingRules } = require('./utils/filter-out-styling-rules');
+
+/**
+ * Extract wepack config from angular-cli 12.2.x
```
🐛  13.x.x

FrozenPandaz commented:
Hey, nice work!

The workaround for `ngcc` is definitely needed and that looks good to me. 

However, I noticed that your changes to `preset-angular-cli` seem to be way further up than I expected. There is logic that used to check if `@angular-devkit/build-angular` is installed and if not has a base config and so forth...

It's down here at this function, that I believe that changes need to be made: 
https://github.com/storybookjs/storybook/blob/next/app/angular/src/server/angular-devkit-build-webpack.ts#L172

For one, there has been a change to the type of `readTsConfig` which is used by the angular preset. This function has been changed from a synchronous function to an async function (returns a Promise) so that needs to be handled.

https://github.com/angular/angular-cli/commit/3e6324f7c7bf774194ff73c848b705e59c052fee#diff-e904a109634cd6284b711e4b36167b489252f899d111a28f8128a2feac3bf245L19

I'll also mention that.. I don't think the `typescript` version in the root `package.json` shouldn't limit us from using `ts` inside of this code. As long as the `typescript` version is able to compile these source files (doesn't have anything to do with Angular), we should be good. What does need to be changed is the `typescript` version that is used within the Angular e2e tests.

ThibaudAV committed:
fix(angular): do best effort to import ngcc

ThibaudAV committed:
feat(angular): split get webpack config in one per supported version

ThibaudAV committed:
fix(angular) fix test for preset angular-cli

ThibaudAV committed:
refactor(angular): remove staticDir for angular-builder because it's deprecated

Use 'staticDirs' Storybook Configuration option instead (main.js)

ThibaudAV committed:
feat(angular): add angularBuilderContext in angular builder

ThibaudAV committed:
feat(angular): add `angularBuilderOptions` to set some options

The goal is to be able to start sb without angularBrowserTarget to use styles
And also to allow nx to set BuilderOptions without `angularBuilderContext`

ThibaudAV commented:
## Breaking Change only for angular 13 :

Storybook is no longer magically based on another configuration of angular.json
Historically `storybook:build` or `defaultProject:build` or env var `STORYBOOK_ANGULAR_PROJECT:build` or ... 
Not having this hidden configuration anymore simplifies the operation. Many issues were opened because this side of the magic was not understood.
A 2nd objective is to promote the use of the storybook builder. This makes storybook compatible for projects with one lib only, one storybook for several applications or one storybook for each application.

To use another configuration we have to use the builder and configure `browserTarget` 

storybook can start without any angular.json configuration (lib only use case, for example)

shilman committed:
Merge branch 'next' into angular/js-way-for-ng13

shilman commented on app/angular/src/builders/start-storybook/index.spec.ts:
```diff
@@ -119,7 +120,6 @@ describe('Start Storybook Builder', () => {
       sslCa: undefined,
       sslCert: undefined,
       sslKey: undefined,
-      staticDir: [],
```
is this a breaking change? we deprecated the `staticDir` CLI option in 6.4 but won't remove it until 7.0

ThibaudAV replied:
I don't think many people use the builder, let alone this option :thinking: 
as this is not the default behaviour nor really a requirement for SB angular
IMHO the builders are still an experiment. But if it works well, we can put them as default (Docs + init cmd) 

and when I was doing my tests, I couldn't find any other way than to delete it to avoid the error :/ 
```
ERR! Error: Conflict when trying to read staticDirs:
ERR! * Storybook's configuration option: 'staticDirs'
ERR! * Storybook's CLI flag: '--staticDir' or '-s'
```

but if necessary and if there is another way to do it we can revert this commit ;)

8thblock commented:
Is this the only way or is there an "official" work around?

frankfajardo commented:
@ThibaudAV question: Am I right in saying that adding/using the storybook builder means there is no need for .storybook/main.js?
